SmartDeviceLink/SDLSystemCapabilityManager.m

```diff
@@ -256,6 +267,139 @@ - (void)updateCapabilityType:(SDLSystemCapabilityType)type completionHandler:(SD
     }
 }
 
+- (nullable SDLWindowCapability *)windowCapabilityWithWindowID:(NSUInteger)windowID {
+    NSArray<SDLDisplayCapability *> *capabilities = self.displays;
+    if (capabilities == nil || capabilities.count == 0) {
+        return nil;
+    }
+    SDLDisplayCapability *display = capabilities[0];
+    for (SDLWindowCapability *windowCapability in display.windowCapabilities) {
+        if (windowCapability.windowID.unsignedIntegerValue == windowID) {
+            return windowCapability;
+        }
+    }
+    return nil;
+}
+
+- (nullable SDLWindowCapability *)defaultMainWindowCapability {
+    return [self windowCapabilityWithWindowID:SDLPredefinedWindowsDefaultWindow];
+}
+
+- (NSArray<SDLDisplayCapability *> *)sdl_createDisplayCapabilityList:(SDLDisplayCapabilities *)display buttons:(NSArray<SDLButtonCapabilities *> *)buttons softButtons:(NSArray<SDLSoftButtonCapabilities *> *)softButtons {
+    // Based on deprecated Display capabilities we don't know if widgets are supported,
+    // The Default MAIN window is the only window we know is supported
+    SDLWindowTypeCapabilities *windowTypeCapabilities = [[SDLWindowTypeCapabilities alloc] initWithType:SDLWindowTypeMain maximumNumberOfWindows:1];
+    
+    SDLDisplayCapability *displayCapability = [[SDLDisplayCapability alloc] initWithDisplayName:(display ? display.displayName: nil)];
+    displayCapability.windowTypeSupported = @[windowTypeCapabilities];
+    
+    // Create a window capability object for the default MAIN window
+    SDLWindowCapability *defaultWindowCapability = [[SDLWindowCapability alloc] init];
+    defaultWindowCapability.windowID = @(SDLPredefinedWindowsDefaultWindow);
+    defaultWindowCapability.buttonCapabilities = buttons.copy;
+    defaultWindowCapability.softButtonCapabilities = softButtons.copy;
+    
+    // return if display capabilities don't exist.
+    if (display == nil) {
+        displayCapability.windowCapabilities = @[defaultWindowCapability];
+        return @[displayCapability];
+    }
+    
+    // copy all available display capabilities
+    defaultWindowCapability.templatesAvailable = display.templatesAvailable.copy;
+    defaultWindowCapability.numCustomPresetsAvailable = display.numCustomPresetsAvailable.copy;
+    defaultWindowCapability.textFields = display.textFields.copy;
+    defaultWindowCapability.imageFields = display.imageFields.copy;
+    
+    if (display.graphicSupported.boolValue) {
+        defaultWindowCapability.imageTypeSupported = @[SDLImageTypeStatic, SDLImageTypeDynamic];
+    } else {
+        defaultWindowCapability.imageTypeSupported = @[SDLImageTypeStatic];
```

Comment:
I don’t know of any information that tells us if static images are supported or not. Reading the mobile API it says weather static or dynamic images are supported on image fields. Because of that we need to imply they exist.

Reply:
Please have a look at https://github.com/smartdevicelink/rpc_spec/blob/5.1.0/MOBILE_API.xml#L2198 which says

> The display's persistent screen supports referencing a static or dynamic image.

Reply:
Right, which would imply that if it's true, then it support either a static image, or a dynamic image, or both. Which would imply that if it's false, the display's persistent screen does not support referencing a static or dynamic image. So then we should not imply that static images are supported, right?

Reply:
Okay got your point. Still this behavior is intentional (and it's proposed exactly this way) due to how AppLink 2.0 works where image support was introduced but our head units with AppLink 2.0 supported only static images. The parameter "graphicSupported" was set to false there to say that dynamic images are not supported. Implied that static images are always supported.. So here's how I read the description:

> The display's persistent screen supports referencing a static (false) or dynamic (true) image.

To migrate to touch screen the suggestion was that developers use "graphicSupported" to use dynamic images in templates or soft buttons. Legacy... which if apps upgrade the lib could break on older head units or at least make them behave extremely slow (as they might try to upload images).

I can solve this issue by looping through "imageFields" array and read "ImageField.imageTypeSupported". This should make conversion to the new capability type crystal clear.

Reply:
Okay, I take your point. I don't think that `imageTypeSupported` will tell us if it supports only static images, unless perhaps the array will be empty? Could you add additional documentation stating that `STATIC` will always be present in the array in backward compatibility situations.